### PR TITLE
Feature/6/modify element controller/element validation

### PR DESCRIPTION
**Pull request by mayankus**

### CODE REVIEW

**Comment by szirbucz** [comment]
@magwas  it seems, there is an error in the model

**Comment by szirbucz** [comment]
You have code coverage problems
```
ERROR: unit coverage miss at com.kodekonveyor.repo.api.ModifyElementController.checkNotFoundServer
ERROR: unit coverage miss at com.kodekonveyor.repo.api.ModifyElementController.checkFoundLocal
ERROR: unit coverage miss at com.kodekonveyor.repo.api.ModifyElementController.checkSyntax
```
You can run coverage test by right click on the project -> coverage as -> JUnit test. Check which lines are not covered and try to update your tests to cover them.

File: .classpath [hunk 1]
```diff
@@ -1,44 +1,38 @@
...
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry excluding="**" kind="src" output="target/test-classes" path="src/test/resources">
+		<attributes>
+			<attribute name="test" value="true"/>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="output" path="target/classes"/>
```
Comment (szirbucz) [inline .classpath]: Revert this file to the state of the develop branch https://stackoverflow.com/questions/33951869/how-to-copy-a-file-from-one-branch-to-another-branch-using-functionality-on-ecli

File: .settings/org.eclipse.jdt.core.prefs [hunk 2]
```diff
@@ -8,8 +8,10 @@ org.eclipse.jdt.core.compiler.debug.lineNumber=generate
 org.eclipse.jdt.core.compiler.debug.localVariable=generate
```
Comment (szirbucz) [inline .settings/org.eclipse.jdt.core.prefs]: revert this file to the state of develop branch (see my comment at .classpath)

File: src/main/java/com/kodekonveyor/repo/api/CheckReferenceResultEnum.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.kodekonveyor.repo.api;
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/CheckReferenceResultEnum.java:2]: Too much new lines

File: src/main/java/com/kodekonveyor/repo/api/CheckReferenceResultEnum.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+public enum CheckReferenceResultEnum {
+  FOUND_LOCAL,
+  FOUND_SERVER,
+  NOT_FOUND_LOCAL,
+  NOT_FOUND_SERVER,
+  REMOTE,
+  SYNTAX
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/CheckReferenceResultEnum.java]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ConstraintException.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.kodekonveyor.repo.api;
+
+public class ConstraintException extends RuntimeException {
+
+  private static final long serialVersionUID = 1L;
+
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ConstraintException.java]: Remove empty lines

File: src/main/java/com/kodekonveyor/repo/api/ElementDTO.java [hunk 6]
```diff
@@ -17,4 +17,5 @@
 
   private String target;
 
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ElementDTO.java:24]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ElementEntity.java [hunk 7]
```diff
@@ -25,4 +25,5 @@
 
   private String target;
 
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ElementEntity.java:29]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ElementModificationDTO.java [hunk 8]
```diff
@@ -9,4 +9,5 @@
 
   private ElementDTO oldElement;
 
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ElementModificationDTO.java]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ElementRepository.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.kodekonveyor.repo.api;
+
+import org.springframework.data.repository.CrudRepository;
+
+public interface ElementRepository extends CrudRepository<ElementEntity, String>{
+
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ElementRepository.java]: Remove 2 empty lines

File: src/main/java/com/kodekonveyor/repo/api/ElementWithRelationsDTO.java [hunk 10]
```diff
@@ -13,4 +13,5 @@
 
   private Set<ElementDTO> targetIn;
 
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ElementWithRelationsDTO.java]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ModificationDTO.java [hunk 11]
```diff
@@ -13,4 +13,5 @@
 
   private Set<ElementModificationDTO> modified;
 
-}
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ModificationDTO.java]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ModifyElementController.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.kodekonveyor.webapp.UrlMapConstants;
+
+@RestController
+public class ModifyElementController {
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ModifyElementController.java]: Remove empty line

File: src/main/java/com/kodekonveyor/repo/api/ModifyElementController.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    return null;
+
+  }
+
+  private void checkReferences(final Optional<ElementEntity> entity) {
+    final ElementEntity elementEntity = entity.get();
+    final CheckReferenceResultEnum sourceReference = checkReferenceService.call(elementEntity, elementEntity.getSource());
+    final CheckReferenceResultEnum targetReference = checkReferenceService.call(elementEntity, elementEntity.getTarget());
+
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/repo/api/ModifyElementController.java]: Remove empty line

File: src/test/java/com/kodekonveyor/exception/ThrowableTester.java [hunk 14]
```diff
@@ -39,10 +39,10 @@ private ThrowableTester _assertThrows(final Thrower thrower) {//NOPMD
   }
 
   public ThrowableTester
-      assertException(final Class<? extends Throwable> klass) {
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/exception/ThrowableTester.java:42]: whitespace only changes.

File: src/test/java/com/kodekonveyor/exception/ThrowableTester.java [hunk 15]
```diff
@@ -12,7 +12,7 @@ public static void assertNoException(final Thrower thrower) {
     try {
       thrower.throwException();
     } catch (final Throwable e) {//NOPMD
-      fail("exception thrown");
+      fail("exception thrown"+ e);
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/exception/ThrowableTester.java]: Change it to `"exception thrown: " + e` (extend with space and colon)

File: src/test/java/com/kodekonveyor/repo/api/ElementDTOTestData.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  public static final String NOT_FOUND_SERVER_ID = "NOT_FOUND_SERVER";
+
+
+
+  public static ElementDTO get() {
+    final ElementDTO element = new ElementDTO();
+    element.setCommit("foundLocalCommit");
+    element.setSource("sourceFoundLocal");
+    element.setTarget("targetFoundLocal");
+    element.setId(FOUND_LOCAL_ID);
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/repo/api/ElementDTOTestData.java]: Code duplication. Create a private method for building ElementDTO
```
private ElementDTO buildElementDTO(
        final String id,
        final String commit,
        final String source,
        final String target
    ) {
    final ElementDTO element = new ElementDTO();

    element.setId(id);
    element.setCommit(commit);
    element.setSource(source);
    element.setTarget(target);

    return element;
}
```
And replace body of each get* method to a call of this method with the right parameters.

File: src/test/java/com/kodekonveyor/repo/api/ElementEntityTestData.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
+package com.kodekonveyor.repo.api;
+
+
+public class ElementEntityTestData {
+
+  public static ElementEntity get() {
+
+    final ElementEntity element = new ElementEntity();
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/repo/api/ElementEntityTestData.java]: Do the same here that you did in ElementDTOTestData to remove the duplications.

File: src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerElementValidationTest.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+@TestedBehaviour("metamodel checking")
+@TestedService("MetamodelCheckService")
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@RunWith(MockitoJUnitRunner.class)
+
+
+public class ModifyElementControllerElementValidationTest extends ModifyElementControllerTestBase {
+
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerElementValidationTest.java]: Too many empty lines

File: src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerTestBase.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+    doReturn(CheckReferenceResultEnum.FOUND_LOCAL).when(checkReferenceService).call(any(),Mockito.eq("sourceFoundLocal"));
+    doReturn(CheckReferenceResultEnum.FOUND_LOCAL).when(checkReferenceService).call(any(),Mockito.eq("targetFoundLocal"));
+    doReturn(CheckReferenceResultEnum.SYNTAX).when(checkReferenceService).call(any(),Mockito.eq("syntax"));
+    doReturn(Optional.ofNullable(ElementEntityTestData.get())).when(elementRepository).findById(ElementDTOTestData.FOUND_LOCAL_ID);
+    doReturn(Optional.ofNullable(ElementEntityTestData.getWithWrongSyntax())).when(elementRepository).findById(ElementDTOTestData.SYNTAX_ID);
+
+    doReturn(CheckReferenceResultEnum.FOUND_SERVER).when(checkReferenceService).call(any(),Mockito.eq("sourceFoundServer"));
+    doReturn(CheckReferenceResultEnum.FOUND_SERVER).when(checkReferenceService).call(any(),Mockito.eq("targetFoundServer"));
+    doReturn(Optional.ofNullable(ElementEntityTestData.getFoundServer())).when(elementRepository).findById(ElementDTOTestData.FOUND_SERVER_ID);
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerTestBase.java]: The above 3 lines are repeated multiple times. Extract it to a method to reduce the code duplication.

File: src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerTestBase.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+    doReturn(CheckReferenceResultEnum.REMOTE).when(checkReferenceService).call(any(),Mockito.eq("sourceRemote"));
```
Comment (szirbucz) [inline src/test/java/com/kodekonveyor/repo/api/ModifyElementControllerTestBase.java]: constants, like "sourceRemote", "targetNotFoundLocal", etc... should come from TestData

**Review by szirbucz — APPROVED** [review]

**Review by szirbucz — CHANGES_REQUESTED** [review]
The former approval was accidental. See the comments

File: .settings/org.eclipse.jdt.core.prefs [hunk 21]
```diff
@@ -703,3 +361,4 @@ org.eclipse.jdt.core.formatter.wrap_before_relational_operator=true
 org.eclipse.jdt.core.formatter.wrap_before_shift_operator=true
 org.eclipse.jdt.core.formatter.wrap_before_string_concatenation=false
 org.eclipse.jdt.core.formatter.wrap_outer_expressions_when_nested=true
+org.eclipse.jdt.core.javaFormatter=org.eclipse.jdt.core.defaultJavaFormatter
```
Comment (szirbucz) [inline .settings/org.eclipse.jdt.core.prefs]: Revert these changes too

**Review by szirbucz — CHANGES_REQUESTED** [review]

**Comment by magwas** [comment]
I have resolved the conflicts for you. Please pull before push.

**Comment by magwas** [comment]
closing due to inactivity

**Pull request closed without merging**